File: web/src/models/stats/dbStats.ts
export type DbHealthStatus = 'healthy' | 'warning' | 'critical' | 'unknown'

export interface IDbTableStats {
  table_name: string
  total_bytes: number
  table_bytes: number
  index_bytes: number
  row_estimate: number
}

export interface IDbStats {
  connected: boolean
  status: DbHealthStatus
  error: string | null
  database_name: string | null
  db_size_bytes: number
  connections_active: number
  connections_idle: number
  connections_idle_in_transaction: number
  connections_total: number
  connections_max: number | null
  cache_hit_ratio: number | null
  pg_stat_statements_enabled: boolean
  slow_query_count: number | null
  deadlocks: number
  temp_bytes: number
  oldest_transaction_age_seconds: number | null
  largest_tables: DbTableStats[]
  checked_at: number | string | null
}

function asObject(v: unknown): Record<string, unknown> {
  return v && typeof v === 'object' && !Array.isArray(v) ? (v as Record<string, unknown>) : {}
}

function asString(v: unknown, fallback = ''): string {
  return typeof v === 'string' ? v : fallback
}

function asNullableString(v: unknown): string | null {
  return typeof v === 'string' && v.length > 0 ? v : null
}

function asNumber(v: unknown, fallback = 0): number {
  if (typeof v === 'number' && Number.isFinite(v)) return v
  if (typeof v === 'string' && v.trim() !== '') {
    const parsed = Number(v)
    if (Number.isFinite(parsed)) return parsed
  }
  return fallback
}

function asNullableNumber(v: unknown): number | null {
  if (v == null) return null
  const parsed = asNumber(v, Number.NaN)
  return Number.isFinite(parsed) ? parsed : null
}

function asBoolean(v: unknown, fallback = false): boolean {
  return typeof v === 'boolean' ? v : fallback
}

function asStatus(v: unknown): DbHealthStatus {
  return v === 'healthy' || v === 'warning' || v === 'critical' ? v : 'unknown'
}

function asCheckedAt(v: unknown): number | string | null {
  if (typeof v === 'number' && Number.isFinite(v)) return v
  if (typeof v === 'string' && v.length > 0) return v
  return null
}

export class DbTableStats implements IDbTableStats {
  table_name = 'unknown'
  total_bytes = 0
  table_bytes = 0
  index_bytes = 0
  row_estimate = 0

  constructor(data?: Partial<IDbTableStats>) {
    if (!data) return
    this.table_name = data.table_name ?? this.table_name
    this.total_bytes = data.total_bytes ?? this.total_bytes
    this.table_bytes = data.table_bytes ?? this.table_bytes
    this.index_bytes = data.index_bytes ?? this.index_bytes
    this.row_estimate = data.row_estimate ?? this.row_estimate
  }

  static from(raw: unknown): DbTableStats {
    const data = asObject(raw)
    return new DbTableStats({
      table_name: asString(data.table_name, 'unknown'),
      total_bytes: asNumber(data.total_bytes),
      table_bytes: asNumber(data.table_bytes),
      index_bytes: asNumber(data.index_bytes),
      row_estimate: asNumber(data.row_estimate),
    })
  }
}

export class DbStats implements IDbStats {
  connected = false
  status: DbHealthStatus = 'unknown'
  error: string | null = null
  database_name: string | null = null
  db_size_bytes = 0
  connections_active = 0
  connections_idle = 0
  connections_idle_in_transaction = 0
  connections_total = 0
  connections_max: number | null = null
  cache_hit_ratio: number | null = null
  pg_stat_statements_enabled = false
  slow_query_count: number | null = null
  deadlocks = 0
  temp_bytes = 0
  oldest_transaction_age_seconds: number | null = null
  largest_tables: DbTableStats[] = []
  checked_at: number | string | null = null

  constructor(data?: Partial<IDbStats>) {
    if (!data) return
    this.connected = data.connected ?? this.connected
    this.status = data.status ?? this.status
    this.error = data.error ?? this.error
    this.database_name = data.database_name ?? this.database_name
    this.db_size_bytes = data.db_size_bytes ?? this.db_size_bytes
    this.connections_active = data.connections_active ?? this.connections_active
    this.connections_idle = data.connections_idle ?? this.connections_idle
    this.connections_idle_in_transaction = data.connections_idle_in_transaction ?? this.connections_idle_in_transaction
    this.connections_total = data.connections_total ?? this.connections_total
    this.connections_max = data.connections_max ?? this.connections_max
    this.cache_hit_ratio = data.cache_hit_ratio ?? this.cache_hit_ratio
    this.pg_stat_statements_enabled = data.pg_stat_statements_enabled ?? this.pg_stat_statements_enabled
    this.slow_query_count = data.slow_query_count ?? this.slow_query_count
    this.deadlocks = data.deadlocks ?? this.deadlocks
    this.temp_bytes = data.temp_bytes ?? this.temp_bytes
    this.oldest_transaction_age_seconds = data.oldest_transaction_age_seconds ?? this.oldest_transaction_age_seconds
    this.largest_tables = data.largest_tables ?? this.largest_tables
    this.checked_at = data.checked_at ?? this.checked_at
  }

  static from(raw: unknown): DbStats {
    const data = asObject(raw)
    const largestTables = Array.isArray(data.largest_tables) ? data.largest_tables.map(DbTableStats.from) : []

    return new DbStats({
      connected: asBoolean(data.connected),
      status: asStatus(data.status),
      error: asNullableString(data.error),
      database_name: asNullableString(data.database_name),
      db_size_bytes: asNumber(data.db_size_bytes),
      connections_active: asNumber(data.connections_active),
      connections_idle: asNumber(data.connections_idle),
      connections_idle_in_transaction: asNumber(data.connections_idle_in_transaction),
      connections_total: asNumber(data.connections_total),
      connections_max: asNullableNumber(data.connections_max),
      cache_hit_ratio: asNullableNumber(data.cache_hit_ratio),
      pg_stat_statements_enabled: asBoolean(data.pg_stat_statements_enabled),
      slow_query_count: asNullableNumber(data.slow_query_count),
      deadlocks: asNumber(data.deadlocks),
      temp_bytes: asNumber(data.temp_bytes),
      oldest_transaction_age_seconds: asNullableNumber(data.oldest_transaction_age_seconds),
      largest_tables: largestTables,
      checked_at: asCheckedAt(data.checked_at),
    })
  }

  connectionUsage(): number | null {
    if (!this.connections_max || this.connections_max <= 0) return null
    return this.connections_total / this.connections_max
  }
}

File: core/src/db/query/stats/DbStats.cpp
#include "db/query/stats/DbStats.hpp"

#include "db/Transactions.hpp"
#include "stats/model/DbStats.hpp"

#include <chrono>
#include <exception>
#include <pqxx/pqxx>

namespace vh::db::query::stats {

namespace {

using vh::stats::model::DbTableStats;

std::uint64_t unixTimestamp() {
    return static_cast<std::uint64_t>(std::chrono::system_clock::to_time_t(std::chrono::system_clock::now()));
}

std::uint64_t asUInt64(const pqxx::row& row, const char* column) {
    const auto field = row[column];
    if (field.is_null()) return 0;
    const auto value = field.as<long long>();
    return value > 0 ? static_cast<std::uint64_t>(value) : 0;
}

std::optional<std::uint64_t> optionalUInt64(const pqxx::row& row, const char* column) {
    const auto field = row[column];
    if (field.is_null()) return std::nullopt;
    const auto value = field.as<long long>();
    if (value < 0) return std::nullopt;
    return static_cast<std::uint64_t>(value);
}

std::optional<double> optionalDouble(const pqxx::row& row, const char* column) {
    const auto field = row[column];
    if (field.is_null()) return std::nullopt;
    return field.as<double>();
}

std::optional<std::string> optionalString(const pqxx::row& row, const char* column) {
    const auto field = row[column];
    if (field.is_null()) return std::nullopt;
    const auto value = field.as<std::string>();
    return value.empty() ? std::nullopt : std::optional<std::string>(value);
}

void applySummary(vh::stats::model::DbStats& stats, const pqxx::result& res) {
    if (res.empty()) return;
    const auto row = res.one_row();
    stats.databaseName = optionalString(row, "database_name");
    stats.dbSizeBytes = asUInt64(row, "db_size_bytes");
    stats.connectionsMax = optionalUInt64(row, "connections_max");
    stats.cacheHitRatio = optionalDouble(row, "cache_hit_ratio");
    stats.deadlocks = asUInt64(row, "deadlocks");
    stats.tempBytes = asUInt64(row, "temp_bytes");
    stats.oldestTransactionAgeSeconds = optionalUInt64(row, "oldest_transaction_age_seconds");
}

void applyConnections(vh::stats::model::DbStats& stats, const pqxx::result& res) {
    for (const auto& row : res) {
        const auto state = row["state"].as<std::string>("");
        const auto count = asUInt64(row, "count");
        stats.connectionsTotal += count;
        if (state == "active") stats.connectionsActive += count;
        else if (state == "idle") stats.connectionsIdle += count;
        else if (state == "idle in transaction") stats.connectionsIdleInTransaction += count;
    }
}

std::vector<DbTableStats> tablesFromResult(const pqxx::result& res) {
    std::vector<DbTableStats> tables;
    tables.reserve(res.size());
    for (const auto& row : res) {
        tables.push_back({
            .tableName = row["table_name"].as<std::string>(""),
            .totalBytes = asUInt64(row, "total_bytes"),
            .tableBytes = asUInt64(row, "table_bytes"),
            .indexBytes = asUInt64(row, "index_bytes"),
            .rowEstimate = asUInt64(row, "row_estimate"),
        });
    }
    return tables;
}

std::string classify(const vh::stats::model::DbStats& stats) {
    if (!stats.connected) return "critical";

    if (stats.connectionsMax && *stats.connectionsMax > 0) {
        const auto ratio = static_cast<double>(stats.connectionsTotal) / static_cast<double>(*stats.connectionsMax);
        if (ratio >= 0.95) return "critical";
        if (ratio >= 0.80) return "warning";
    }

    if (stats.oldestTransactionAgeSeconds && *stats.oldestTransactionAgeSeconds >= 86400) return "critical";
    if (stats.deadlocks > 0) return "warning";
    if (stats.oldestTransactionAgeSeconds && *stats.oldestTransactionAgeSeconds >= 3600) return "warning";
    if (stats.cacheHitRatio && *stats.cacheHitRatio > 0.0 && *stats.cacheHitRatio < 0.80) return "warning";

    return "healthy";
}

std::optional<std::uint64_t> slowQueryCount() {
    try {
        return Transactions::exec("DbStats::slowQueryCount", [](pqxx::work& txn) -> std::optional<std::uint64_t> {
            const auto res = txn.exec(R"SQL(
                SELECT COUNT(*) AS count
                FROM pg_stat_statements
                WHERE calls > 0
                  AND mean_exec_time >= 1000.0;
            )SQL");

            if (res.empty()) return 0;
            return asUInt64(res.one_row(), "count");
        });
    } catch (const std::exception&) {
        return std::nullopt;
    }
}

}

std::shared_ptr<vh::stats::model::DbStats> DbStats::snapshot() {
    auto stats = std::make_shared<vh::stats::model::DbStats>();
    stats->checkedAt = unixTimestamp();

    try {
        Transactions::exec("DbStats::snapshot", [&](pqxx::work& txn) {
            stats->connected = true;
            applySummary(*stats, txn.exec(pqxx::prepped{"db_stats.summary"}));
            applyConnections(*stats, txn.exec(pqxx::prepped{"db_stats.connections"}));
            const auto extension = txn.exec(pqxx::prepped{"db_stats.pg_stat_statements_enabled"});
            stats->pgStatStatementsEnabled = !extension.empty() && extension.one_row()["enabled"].as<bool>(false);
            stats->largestTables = tablesFromResult(txn.exec(pqxx::prepped{"db_stats.largest_tables"}));
        });

        if (stats->pgStatStatementsEnabled) stats->slowQueryCount = slowQueryCount();
    } catch (const std::exception& e) {
        stats->connected = false;
        stats->status = "critical";
        stats->error = e.what();
        return stats;
    }

    stats->status = classify(*stats);
    return stats;
}

}
